clang/include/clang/Basic/UnsignedOrNone.h

```diff
@@ -0,0 +1,53 @@
+//===- UnsignedOrNone.h - simple optional index-----*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+/// \file
+/// Defines clang::UnsignedOrNone.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_BASIC_UNSIGNED_OR_NONE_H
+#define LLVM_CLANG_BASIC_UNSIGNED_OR_NONE_H
+
+#include <cassert>
+#include <optional>
+
+namespace clang {
+
+struct UnsignedOrNone {
+  constexpr UnsignedOrNone(std::nullopt_t) : Rep(0) {}
```

Comment:
I might suggest using `std::numeric_limits<unsigned>::max()` as the sentinel value as it probably makes the 'rep' easier to read in a debugger, and prevents the adjustment in the 2 places.

Reply:
Reading on a debugger is solvable, most debuggers support some sort of description of a user type which can be used to improve how it is presented.

I am afraid of the change to UINT_MAX as sentinel because of the translation back and forward to unsigned types of different sizes, such as bitfields.

And this is mostly how we already serialize them anyway.